recirq/qml_lfe/learn_states_q.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2021 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Draw I + P pauli product data using two copy strategy.
+style circuits.
+
+learn_states_q.py --n=6 --n_paulis=20 --batch_size=500 --n_shots=1000
+
+Will create paulistring circuits all on 12 qubits (two systems of size 6).
+Once the circuits are generated, batch_size sweeps and a single circuit at
+a timewill be sent will be sent for simulation/execution, drawing n_shots
+total samples from each one using `run_sweep` in Cirq. By default the
+bitstring data will be saved in the data folder. One can also set
+`use_engine` to True in order to run this against a processor on
+quantum engine.
+"""
+from typing import Dict, List
+
+import os
+import cirq
+import numpy as np
+import sympy
+from . import circuit_blocks
+from . import run_config
+
+from absl import app
+from absl import logging
+
+
+def _create_randomized_sweeps(
+    hidden_p: str, symbols: Dict[str, int], num_reps: int
+) -> List[Dict[str, int]]:
+    last_i = 0
+    for i, pauli in enumerate(hidden_p):
+        if pauli != "I":
+            last_i = i
+
+    sign_p = -1 if np.random.random() < 0.5 else 1
+    all_sweeps = []
+    for _ in range(num_reps):
+        current_sweep = dict()
+        for twocopy in [0, 1]:
+            parity = sign_p * (1 if np.random.random() <= 0.95 else -1)
```

Comment:
What is the significance of 0.95 here?

Reply:
This is a detail in the paper. We are actually generating rho = I +- 0.95 P (details in section A.2.a) https://arxiv.org/pdf/2112.00778.pdf

Reply:
in general, comments referring to a specific section in the paper are highly encouraged

Reply:
ah, I see this is in the docstring.

Reply:
This, and above might read better if you 

1) allow optional passing of a numpy randomstate like in cirq for optional determinism
2) use `rs.choice([1,-1], p=[0.95, 0.05])`

Reply:
especially since you use `choice` later

Reply:
Changed this to make use of randomstate. I was a little lax with integrating this in some other places in the code so I may circle back and improve later.